internal/controllers/rollout/synthesizer.go

```diff
@@ -0,0 +1,127 @@
+package rollout
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+)
+
+type synthController struct {
+	client client.Client
+}
+
+func NewSynthesizerController(mgr ctrl.Manager) error {
+	c := &synthController{
+		client: mgr.GetClient(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv1.Synthesizer{}).
+		Watches(&apiv1.Composition{}, newCompositionHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "synthesizerRolloutController")).
+		Complete(c)
+}
+
+func (c *synthController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	syn := &apiv1.Synthesizer{}
+	err := c.client.Get(ctx, req.NamespacedName, syn)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("gettting synthesizer: %w", err))
+	}
+	logger = logger.WithValues("synthesizerName", syn.Name, "synthesizerNamespace", syn.Namespace, "synthesizerGeneration", syn.Generation)
+
+	compList := &apiv1.CompositionList{}
+	err = c.client.List(ctx, compList, client.MatchingFields{
+		manager.IdxCompositionsBySynthesizer: syn.Name,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing compositions: %w", err)
+	}
+
+	// randomize list to avoid always rolling out changes in the same order
+	rand.Shuffle(len(compList.Items), func(i, j int) { compList.Items[i] = compList.Items[j] })
+
+	for _, comp := range compList.Items {
+		comp := comp
+		logger := logger.WithValues("compositionName", comp.Name, "compositionNamespace", comp.Namespace, "compositionGeneration", comp.Generation)
+
+		// Compositions aren't eligible to receive an updated synthesizer when:
+		// - They haven't ever been synthesized (they'll use the latest inputs anyway)
+		// - They are currently being synthesized or deleted
```

Comment:
The current synthesis could be using the previous generation of the synthesizer, would it not get stuck there if there are no composition changes until the next synthesizer release?

Reply:
The synthesized time and/or uuid will change, which will cause this code to be executed again. So we'll always catch up eventually.